File: app/src/main/java/org/wikipedia/gallery/GalleryThumbnailScrollView.java
package org.wikipedia.gallery;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.wikipedia.R;
import org.wikipedia.dataclient.mwapi.MwQueryPage;
import org.wikipedia.util.StringUtil;
import org.wikipedia.views.ViewUtil;

import java.util.List;

public class GalleryThumbnailScrollView extends RecyclerView {
    @NonNull private final Animation mPressAnimation;
    @NonNull private final Animation mReleaseAnimation;
    @Nullable private GalleryViewListener mListener;

    public GalleryThumbnailScrollView(Context context) {
        this(context, null);
    }

    public GalleryThumbnailScrollView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public GalleryThumbnailScrollView(@NonNull Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));

        mPressAnimation = AnimationUtils.loadAnimation(context, R.anim.thumbnail_item_press);
        mReleaseAnimation = AnimationUtils.loadAnimation(context, R.anim.thumbnail_item_release);
    }

    public interface GalleryViewListener {
        void onGalleryItemClicked(@NonNull ImageView view, @NonNull String thumbUrl, @NonNull String imageName);
    }

    public void setGalleryViewListener(@Nullable GalleryViewListener listener) {
        mListener = listener;
    }

    public void setGalleryList(@NonNull List<MwQueryPage> list) {
        setAdapter(new GalleryViewAdapter(list));
    }

    private class GalleryItemHolder extends ViewHolder implements OnClickListener, OnTouchListener {
        private final ImageView mImageView;
        private MwQueryPage mGalleryItem;

        GalleryItemHolder(View itemView) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.gallery_thumbnail_image);
        }

        public void bindItem(MwQueryPage item) {
            mGalleryItem = item;
            mImageView.setFocusable(true);
            mImageView.setOnClickListener(this);
            mImageView.setOnTouchListener(this);
            ViewUtil.loadImage(mImageView, mGalleryItem.imageInfo().getThumbUrl());
        }

        @Override
        public void onClick(View v) {
            if (mListener != null) {
                mListener.onGalleryItemClicked((ImageView)v, mGalleryItem.imageInfo().getThumbUrl(), StringUtil.addUnderscores(mGalleryItem.title()));
            }
        }

        @Override
        public boolean onTouch(View v, MotionEvent event) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    v.startAnimation(mPressAnimation);
                    break;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                    v.startAnimation(mReleaseAnimation);
                    break;
                default:
                    break;
            }
            return false;
        }
    }

    private final class GalleryViewAdapter extends RecyclerView.Adapter<GalleryItemHolder> {
        @NonNull private final List<MwQueryPage> list;

        GalleryViewAdapter(@NonNull List<MwQueryPage> list) {
            this.list = list;
        }

        @Override
        public int getItemCount() {
            return  list.size();
        }

        @Override
        public GalleryItemHolder onCreateViewHolder(ViewGroup parent, int pos) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_gallery_thumbnail, parent, false);
            return new GalleryItemHolder(view);
        }

        @Override
        public void onBindViewHolder(GalleryItemHolder holder, int pos) {
            holder.bindItem(list.get(pos));
        }
    }
}

File: app/src/main/java/org/wikipedia/gallery/GalleryThumbnailScrollView.kt
package org.wikipedia.gallery

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ImageView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.wikipedia.R
import org.wikipedia.dataclient.mwapi.MwQueryPage
import org.wikipedia.util.StringUtil
import org.wikipedia.views.ViewUtil.loadImage

class GalleryThumbnailScrollView constructor(context: Context, attrs: AttributeSet? = null) : RecyclerView(context, attrs) {

    fun interface GalleryViewListener {
        fun onGalleryItemClicked(view: ImageView, thumbUrl: String, imageName: String)
    }

    private val pressAnimation = AnimationUtils.loadAnimation(context, R.anim.thumbnail_item_press)
    private val releaseAnimation = AnimationUtils.loadAnimation(context, R.anim.thumbnail_item_release)
    var listener: GalleryViewListener? = null

    init {
        layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
    }

    fun setGalleryList(list: List<MwQueryPage>) {
        adapter = GalleryViewAdapter(list)
    }

    @SuppressLint("ClickableViewAccessibility")
    private inner class GalleryItemHolder constructor(itemView: View) : ViewHolder(itemView), OnClickListener, OnTouchListener {
        private lateinit var galleryItem: MwQueryPage
        private val imageView = itemView.findViewById<ImageView>(R.id.gallery_thumbnail_image)

        fun bindItem(item: MwQueryPage) {
            galleryItem = item
            galleryItem.imageInfo()?.let {
                imageView.isFocusable = true
                imageView.setOnClickListener(this)
                imageView.setOnTouchListener(this)
                loadImage(imageView, it.thumbUrl)
            }
        }

        override fun onClick(v: View) {
            galleryItem.imageInfo()?.let {
                listener?.onGalleryItemClicked(v as ImageView, it.thumbUrl, StringUtil.addUnderscores(galleryItem.title()))
            }
        }

        override fun onTouch(v: View, event: MotionEvent): Boolean {
            when (event.action) {
                MotionEvent.ACTION_DOWN -> v.startAnimation(pressAnimation)
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> v.startAnimation(releaseAnimation)
            }
            return false
        }
    }

    private inner class GalleryViewAdapter constructor(private val list: List<MwQueryPage>) : Adapter<GalleryItemHolder>() {
        override fun getItemCount(): Int {
            return list.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, pos: Int): GalleryItemHolder {
            return GalleryItemHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.item_gallery_thumbnail, parent, false))
        }

        override fun onBindViewHolder(holder: GalleryItemHolder, pos: Int) {
            holder.bindItem(list[pos])
        }
    }
}
